src/org/daisy/dotify/formatter/impl/page/PageSequenceBuilder2.java

```diff
@@ -686,9 +699,45 @@ private static float height(List<RowGroup> rg, boolean useLastUnitSize) {
         }
     }
 
-    private void addRows(List<RowGroup> head, PageImpl p) {
+    private BlockContext addRows(List<RowGroup> head, PageImpl p, BlockContext blockContext) {
         int i = head.size();
         for (RowGroup rg : head) {
+
+            /*
+                At this point we expect keep="page" is set when the condition evaluates to false, which means that
+                each block should not be spanning multiple pages. This is required so we don't print just a part
+                of the block when the display-when attribute is set to false.
+                We expect display-when is either set to "true" (or missing, which is the same) or
+                "(! $starts-at-top-of-page)". Other values are not permitted.
+                These restrictions are added in the OBFL Parser and will lead to exceptions being thrown.
+
+                Above assumption deserves some more explanation for a good understanding:
+                we need it because we evaluate display-when for each RowGroup while normally
+                it should be evaluated _only once per block.
+
+                This is fine in the two mentioned cases:
+                    - display-when="true": trivial.
+                    - display-when="(! $starts-at-top-of-page)":
+                        - if this evaluates to true, it means the page must already contain a RowImpl(*),
+                          so regardless of whether the current RowGroup belongs to the same block or a new
+                          block, it must be rendered.
+                        - if it evaluates to false, it means the page does not already contain a RowImpl,
+                          so we know it must be the start of a new page and block (because keep="page" in
+                          this case).
+                (*) Refer to the line below where you set .topOfPage(false). The line should however be moved
```

Comment:
Which line below?
Who is "you"?

Reply:
The line below where we set `.topOfPage` if you look at the full file it's line 737 or something. Have changed you to we as it's less distinct. We the implementors.